File: packages/queries/chats.ts
import type {
  Chat,
  ChatWithMessages,
  CreateChatRequest,
  ExecutionProcess,
  UpdateChatRequest,
} from '@openflow/generated';
import { createLogger } from '@openflow/utils';
import { createChatSchema, updateChatSchema } from '@openflow/validation';
import { invoke } from './utils.js';

/**
 * Logger for chat query operations.
 * Logs at DEBUG level for query calls, INFO for successes, ERROR for failures.
 */
const logger = createLogger('queries:chats');

/**
 * Chat query wrappers for Tauri IPC.
 * Thin wrappers around invoke() calls with type safety.
 * Input validation is performed using Zod schemas before invoking Tauri commands.
 *
 * All functions include:
 * - Try/catch error handling with re-throw for React Query
 * - Logging at appropriate levels (DEBUG on call, INFO on success, ERROR on failure)
 * - Input validation via Zod schemas for mutations
 *
 * @example
 * ```ts
 * // List all chats for a task
 * const chats = await chatQueries.list('task-123');
 *
 * // Get a single chat with messages
 * const chat = await chatQueries.get('chat-456');
 *
 * // Create a new chat
 * const newChat = await chatQueries.create({
 *   projectId: 'project-123',
 *   title: 'New Discussion',
 * });
 *
 * // Archive a chat
 * await chatQueries.archive('chat-456');
 * ```
 */
export const chatQueries = {
  /**
   * List all chats for a task.
   * @param taskId - Task ID to filter chats
   * @returns Promise resolving to array of chats
   * @throws Error if the query fails (re-thrown for React Query)
   */
  list: async (taskId: string): Promise<Chat[]> => {
    logger.debug('Listing chats for task', { taskId });

    try {
      const chats = await invoke<Chat[]>('list_chats', { taskId });

      logger.info('Chats listed successfully', {
        taskId,
        count: chats.length,
        titles: chats.slice(0, 3).map((c) => c.title),
        hasMore: chats.length > 3,
      });

      return chats;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to list chats', {
        taskId,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * List standalone chats for a project (chats without a task).
   * @param projectId - Project ID to filter chats
   * @returns Promise resolving to array of standalone chats
   * @throws Error if the query fails (re-thrown for React Query)
   */
  listStandalone: async (projectId: string): Promise<Chat[]> => {
    logger.debug('Listing standalone chats for project', { projectId });

    try {
      const chats = await invoke<Chat[]>('list_standalone_chats', { projectId });

      logger.info('Standalone chats listed successfully', {
        projectId,
        count: chats.length,
        titles: chats.slice(0, 3).map((c) => c.title),
        hasMore: chats.length > 3,
      });

      return chats;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to list standalone chats', {
        projectId,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * List all chats for a project (both task-linked and standalone).
   * @param projectId - Project ID to filter chats
   * @returns Promise resolving to array of all project chats
   * @throws Error if the query fails (re-thrown for React Query)
   */
  listByProject: async (projectId: string): Promise<Chat[]> => {
    logger.debug('Listing all chats for project', { projectId });

    try {
      const chats = await invoke<Chat[]>('list_chats_by_project', { projectId });

      logger.info('Project chats listed successfully', {
        projectId,
        count: chats.length,
        titles: chats.slice(0, 3).map((c) => c.title),
        hasMore: chats.length > 3,
      });

      return chats;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to list chats by project', {
        projectId,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Get a single chat by ID with its messages.
   * @param id - Chat ID
   * @returns Promise resolving to chat with messages
   * @throws Error if the query fails (re-thrown for React Query)
   */
  get: async (id: string): Promise<ChatWithMessages> => {
    logger.debug('Getting chat', { id });

    try {
      const result = await invoke<ChatWithMessages>('get_chat', { id });

      logger.info('Chat retrieved successfully', {
        id,
        title: result.title,
        messageCount: result.messages.length,
        hasTask: !!result.taskId,
      });

      return result;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to get chat', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Create a new chat.
   * Input is validated against createChatSchema before invoking.
   * @param request - Chat creation request
   * @returns Promise resolving to the created chat
   * @throws Error if validation or query fails (re-thrown for React Query)
   */
  create: async (request: CreateChatRequest): Promise<Chat> => {
    logger.debug('Creating chat', {
      projectId: request.projectId,
      taskId: request.taskId,
      hasTitle: !!request.title,
      hasWorkflowStep: !!request.workflowStepIndex,
    });

    try {
      const validated = createChatSchema.parse(request);
      const chat = await invoke<Chat>('create_chat', { request: validated });

      logger.info('Chat created successfully', {
        id: chat.id,
        title: chat.title,
        projectId: chat.projectId,
        taskId: chat.taskId,
      });

      return chat;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to create chat', {
        projectId: request.projectId,
        taskId: request.taskId,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Update an existing chat.
   * Input is validated against updateChatSchema before invoking.
   * @param id - Chat ID to update
   * @param request - Update request with fields to change
   * @returns Promise resolving to the updated chat
   * @throws Error if validation or query fails (re-thrown for React Query)
   */
  update: async (id: string, request: UpdateChatRequest): Promise<Chat> => {
    logger.debug('Updating chat', {
      id,
      hasTitle: 'title' in request,
      hasSetupCompleted: 'setupCompletedAt' in request,
    });

    try {
      const validated = updateChatSchema.parse(request);
      const chat = await invoke<Chat>('update_chat', { id, request: validated });

      logger.info('Chat updated successfully', {
        id: chat.id,
        title: chat.title,
      });

      return chat;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to update chat', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Delete a chat by ID.
   * This will cascade delete all associated messages.
   * @param id - Chat ID to delete
   * @returns Promise resolving when deletion is complete
   * @throws Error if the query fails (re-thrown for React Query)
   */
  delete: async (id: string): Promise<void> => {
    logger.debug('Deleting chat', { id });

    try {
      await invoke<void>('delete_chat', { id });

      logger.info('Chat deleted successfully', { id });
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to delete chat', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Archive a chat by ID.
   * Archived chats are hidden from list queries.
   * @param id - Chat ID to archive
   * @returns Promise resolving to the archived chat
   * @throws Error if the query fails (re-thrown for React Query)
   */
  archive: async (id: string): Promise<Chat> => {
    logger.debug('Archiving chat', { id });

    try {
      const chat = await invoke<Chat>('archive_chat', { id });

      logger.info('Chat archived successfully', {
        id: chat.id,
        title: chat.title,
      });

      return chat;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to archive chat', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Unarchive a chat by ID.
   * Makes the chat visible in list queries again.
   * @param id - Chat ID to unarchive
   * @returns Promise resolving to the unarchived chat
   * @throws Error if the query fails (re-thrown for React Query)
   */
  unarchive: async (id: string): Promise<Chat> => {
    logger.debug('Unarchiving chat', { id });

    try {
      const chat = await invoke<Chat>('unarchive_chat', { id });

      logger.info('Chat unarchived successfully', {
        id: chat.id,
        title: chat.title,
      });

      return chat;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to unarchive chat', {
        id,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * List all archived chats across all projects.
   * @returns Promise resolving to array of archived chats
   * @throws Error if the query fails (re-thrown for React Query)
   */
  listArchived: async (): Promise<Chat[]> => {
    logger.debug('Listing archived chats');

    try {
      const chats = await invoke<Chat[]>('list_archived_chats');

      logger.info('Archived chats listed successfully', {
        count: chats.length,
        titles: chats.slice(0, 3).map((c) => c.title),
        hasMore: chats.length > 3,
      });

      return chats;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to list archived chats', {
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Start a workflow step execution for a chat.
   * This triggers the executor to run on the chat's initial prompt.
   * @param chatId - The chat ID to start the workflow step for
   * @returns Promise resolving to the execution process
   * @throws Error if the query fails (re-thrown for React Query)
   */
  startWorkflowStep: async (chatId: string): Promise<ExecutionProcess> => {
    logger.debug('Starting workflow step', { chatId });

    try {
      const process = await invoke<ExecutionProcess>('start_workflow_step', { chatId });

      logger.info('Workflow step started successfully', {
        chatId,
        processId: process.id,
        status: process.status,
      });

      return process;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to start workflow step', {
        chatId,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Toggle the completion status of a workflow step (chat).
   * If the step is incomplete, it will be marked as complete.
   * If the step is complete, it will be marked as incomplete.
   * @param chatId - The chat ID to toggle completion for
   * @returns Promise resolving to the updated chat
   * @throws Error if the query fails (re-thrown for React Query)
   */
  toggleStepComplete: async (chatId: string): Promise<Chat> => {
    logger.debug('Toggling step completion', { chatId });

    try {
      const chat = await invoke<Chat>('toggle_step_complete', { chatId });

      logger.info('Step completion toggled successfully', {
        chatId,
        isComplete: !!chat.setupCompletedAt,
      });

      return chat;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to toggle step completion', {
        chatId,
        error: errorMessage,
      });
      throw error;
    }
  },
};

File: src-tauri/src/commands/chats.rs
//! Tauri commands for chat operations.
//!
//! These commands provide the IPC interface for chat CRUD operations.
//! Each command is a thin wrapper around ChatService methods.

use tauri::State;

use crate::commands::AppState;
use openflow_contracts::{
    Chat, ChatWithMessages, CreateChatRequest, ExecutionProcess, UpdateChatRequest,
};
use openflow_core::services::chat;

/// List chats for a task.
///
/// Returns chats ordered by workflow_step_index ASC, then created_at ASC.
///
/// # Arguments
/// * `task_id` - The task to list chats for
#[tauri::command]
pub async fn list_chats(state: State<'_, AppState>, task_id: String) -> Result<Vec<Chat>, String> {
    let pool = state.db.lock().await;
    chat::list(&pool, &task_id).await.map_err(|e| e.to_string())
}

/// Get a single chat by ID with its associated messages.
///
/// Returns the chat with all its messages if found, or an error if not found.
#[tauri::command]
pub async fn get_chat(state: State<'_, AppState>, id: String) -> Result<ChatWithMessages, String> {
    let pool = state.db.lock().await;
    chat::get(&pool, &id).await.map_err(|e| e.to_string())
}

/// Create a new chat.
///
/// Returns the newly created chat with generated ID and timestamps.
#[tauri::command]
pub async fn create_chat(
    state: State<'_, AppState>,
    request: CreateChatRequest,
) -> Result<Chat, String> {
    let pool = state.db.lock().await;
    chat::create(&pool, request)
        .await
        .map_err(|e| e.to_string())
}

/// Update an existing chat.
///
/// Only the provided fields will be updated. Returns the updated chat.
#[tauri::command]
pub async fn update_chat(
    state: State<'_, AppState>,
    id: String,
    request: UpdateChatRequest,
) -> Result<Chat, String> {
    let pool = state.db.lock().await;
    chat::update(&pool, &id, request)
        .await
        .map_err(|e| e.to_string())
}

/// Delete a chat by ID.
///
/// This will cascade delete all associated messages.
#[tauri::command]
pub async fn delete_chat(state: State<'_, AppState>, id: String) -> Result<(), String> {
    let pool = state.db.lock().await;
    chat::delete(&pool, &id).await.map_err(|e| e.to_string())
}

/// Archive a chat by ID.
///
/// Sets the archived_at timestamp. Archived chats are hidden from list queries.
#[tauri::command]
pub async fn archive_chat(state: State<'_, AppState>, id: String) -> Result<Chat, String> {
    let pool = state.db.lock().await;
    chat::archive(&pool, &id).await.map_err(|e| e.to_string())
}

/// Unarchive a chat by ID.
///
/// Clears the archived_at timestamp, making the chat visible again.
#[tauri::command]
pub async fn unarchive_chat(state: State<'_, AppState>, id: String) -> Result<Chat, String> {
    let pool = state.db.lock().await;
    chat::unarchive(&pool, &id).await.map_err(|e| e.to_string())
}

/// List standalone chats for a project.
///
/// Returns chats that are not associated with any task (task_id IS NULL).
/// Ordered by created_at DESC (most recent first).
///
/// # Arguments
/// * `project_id` - The project to list standalone chats for
#[tauri::command]
pub async fn list_standalone_chats(
    state: State<'_, AppState>,
    project_id: String,
) -> Result<Vec<Chat>, String> {
    let pool = state.db.lock().await;
    chat::list_standalone(&pool, &project_id)
        .await
        .map_err(|e| e.to_string())
}

/// List all chats for a project.
///
/// Includes both task-linked chats and standalone chats.
/// Ordered by created_at DESC.
///
/// # Arguments
/// * `project_id` - The project to list all chats for
#[tauri::command]
pub async fn list_chats_by_project(
    state: State<'_, AppState>,
    project_id: String,
) -> Result<Vec<Chat>, String> {
    let pool = state.db.lock().await;
    chat::list_by_project(&pool, &project_id)
        .await
        .map_err(|e| e.to_string())
}

/// List all archived chats across all projects.
///
/// Returns archived chats ordered by archived_at DESC (most recently archived first).
#[tauri::command]
pub async fn list_archived_chats(state: State<'_, AppState>) -> Result<Vec<Chat>, String> {
    let pool = state.db.lock().await;
    chat::list_archived(&pool).await.map_err(|e| e.to_string())
}

/// Toggle the completion status of a workflow step (chat).
///
/// If the step is incomplete (setup_completed_at is None), marks it as complete.
/// If the step is complete, marks it as incomplete.
///
/// # Arguments
/// * `id` - The chat ID to toggle completion for
///
/// # Returns
/// The updated Chat with toggled completion status.
#[tauri::command]
pub async fn toggle_step_complete(state: State<'_, AppState>, id: String) -> Result<Chat, String> {
    let pool = state.db.lock().await;
    chat::toggle_step_complete(&pool, &id)
        .await
        .map_err(|e| e.to_string())
}

/// Start a workflow step execution for a chat.
///
/// This triggers the executor to run on the chat's initial prompt.
/// Creates a new execution process and begins the AI agent execution.
///
/// # Arguments
/// * `chat_id` - The chat ID to start the workflow step for
///
/// # Note
/// This command requires ProcessService to be fully implemented.
/// Currently returns an error indicating the feature is not yet available.
#[tauri::command]
pub async fn start_workflow_step(
    _state: State<'_, AppState>,
    _chat_id: String,
) -> Result<ExecutionProcess, String> {
    // TODO: Implement once ProcessService is complete (Phase 13)
    // The implementation will:
    // 1. Get the chat and verify it exists
    // 2. Get the executor profile for the chat
    // 3. Create a new ExecutionProcess record
    // 4. Start the executor process (CLI tool)
    // 5. Return the process record
    Err(
        "Workflow step execution not yet implemented. See Phase 13 of the implementation plan."
            .to_string(),
    )
}
